streaming/src/queue/queue_service.h

```diff
@@ -0,0 +1,192 @@
+  void QueueThreadCallback() { queue_service_.run(); }
+
+ protected:
+  /// CoreWorker C++ pointer of current actor
+  CoreWorker *core_worker_;
+  /// actor_id actor id of current actor
+  ActorID actor_id_;
+  /// Helper function, parse message buffer to Message object.
+  std::shared_ptr<Message> ParseMessage(std::shared_ptr<LocalMemoryBuffer> buffer);
+
+ private:
+  /// Map from queue id to a actor id of the queue's peer actor.
+  std::unordered_map<ObjectID, ActorID> actors_;
+  /// Map from queue id to a transport of the queue's peer actor.
+  std::unordered_map<ObjectID, std::shared_ptr<Transport>> out_transports_;
+  /// The internal thread which asio service run with.
+  std::thread queue_thread_;
+  /// The internal asio service.
+  boost::asio::io_service queue_service_;
+  /// The asio work which keeps queue_service_ alive.
+  boost::asio::io_service::work queue_dummy_work_;
+};
+
+/// UpstreamService holds and manages all upstream queues of current actor.
+class UpstreamService : public QueueService {
+ public:
+  /// Construct a UpstreamService instance.
+  UpstreamService(CoreWorker *core_worker, const ActorID &actor_id)
+      : QueueService(core_worker, actor_id) {}
+  /// Create a upstream queue.
+  /// \param[in] queue_id queue id of the queue to be created.
+  /// \param[in] peer_actor_id actor id of peer actor.
+  /// \param[in] size the max memory size of the queue.
+  std::shared_ptr<WriterQueue> CreateUpstreamQueue(const ObjectID &queue_id,
+                                                   const ActorID &peer_actor_id,
+                                                   uint64_t size);
+  /// Check whether the upstream queue specified by queue_id exists or not.
+  bool UpstreamQueueExists(const ObjectID &queue_id);
+  /// Wait all queues in queue_ids vector ready, until timeout.
+  /// \param[in] queue_ids a group of queues.
+  /// \param[in] timeout_ms max timeout time interval for wait all queues.
+  /// \param[out] failed_queues a group of queues which are not ready when timeout.
+  void WaitQueues(const std::vector<ObjectID> &queue_ids, int64_t timeout_ms,
+                  std::vector<ObjectID> &failed_queues);
+  /// Handle notify message from corresponded downstream queue.
+  void OnNotify(std::shared_ptr<NotificationMessage> notify_msg);
+  /// Obtain upstream queue specified by queue_id.
+  std::shared_ptr<streaming::WriterQueue> GetUpQueue(const ObjectID &queue_id);
+  /// Release all upstream queues
+  void ReleaseAllUpQueues();
+
+  virtual void DispatchMessageInternal(
+      std::shared_ptr<LocalMemoryBuffer> buffer,
+      std::function<void(std::shared_ptr<LocalMemoryBuffer>)> callback) override;
+
+  static std::shared_ptr<UpstreamService> CreateService(CoreWorker *core_worker,
+                                                        const ActorID &actor_id);
+  static std::shared_ptr<UpstreamService> GetService();
+
+  static RayFunction PEER_SYNC_FUNCTION;
```

Comment:
static variables should use similar variable names like others, e.g. `peer_sync_function_`

Reply:
> static variables should use similar variable names like others, e.g. `peer_sync_function_`

fixed.